src/Workspaces/Remote/ServiceHub/Shared/RemoteEndPoint.cs

```diff
@@ -173,6 +188,98 @@ public async Task<T> InvokeAsync<T>(string targetName, IReadOnlyList<object?> ar
             }
         }
 
+        public static async Task WriteDataToNamedPipeAsync<TData>(string pipeName, TData data, Func<ObjectWriter, TData, CancellationToken, Task> dataWriter, CancellationToken cancellationToken)
+        {
+            const int BufferSize = 4 * 1024;
+
+            try
+            {
+                var pipe = new NamedPipeClientStream(serverName: ".", pipeName, PipeDirection.Out);
+
+                bool success = false;
+                try
+                {
+                    await ConnectPipeAsync(pipe, cancellationToken).ConfigureAwait(false);
+                    success = true;
+                }
+                finally
+                {
+                    if (!success)
+                    {
+                        pipe.Dispose();
+                    }
+                }
+
+                // Transfer ownership of the pipe to BufferedStream, it will dispose it:
+                using var stream = new BufferedStream(pipe, BufferSize);
+
+                using (var objectWriter = new ObjectWriter(stream, leaveOpen: true, cancellationToken))
+                {
+                    await dataWriter(objectWriter, data, cancellationToken).ConfigureAwait(false);
+                }
+
+                await stream.FlushAsync(cancellationToken).ConfigureAwait(false);
+            }
+            catch (Exception) when (cancellationToken.IsCancellationRequested)
+            {
+                // The stream has closed before we had chance to check cancellation.
+                cancellationToken.ThrowIfCancellationRequested();
+            }
+        }
+
+        private static async Task ConnectPipeAsync(NamedPipeClientStream pipe, CancellationToken cancellationToken)
+        {
+            const int ConnectWithoutTimeout = 1;
+            const int MaxRetryAttemptsForFileNotFoundException = 3;
```

Comment:
when does this happen?

Reply:
IDK. This will be deleted once we move to IServiceBroker, so I'm not really keen on figuring that out.